mce-fbdev.c

```diff
@@ -0,0 +1,281 @@
+ */
+bool mce_fbdev_is_open(void)
+{
+    return mce_fbdev_handle != -1;
+}
+
+/** Open frame buffer device unless denied
+ *
+ * @return true on success, or false on failure
+ */
+bool mce_fbdev_open(void)
+{
+#ifdef ENABLE_HYBRIS
+    if( fbdev_use_hybris )
+        goto EXIT;
+#endif
+
+    if( mce_fbdev_handle != -1 )
+        goto EXIT;
+
+    if( !mce_fbdev_allow_open )
+        goto EXIT;
+
+    mce_log(LL_NOTICE, "open frame buffer device");
+
+    if( (mce_fbdev_handle = open(FB_DEVICE, O_RDWR)) == -1 ) {
+        if( errno != ENOENT )
+            mce_log(LL_WARN, "failed to open frame buffer device: %m");
+        goto EXIT;
+    }
+
+    mce_log(LL_DEBUG, "frame buffer device opened");
+
+EXIT:
+    return mce_fbdev_handle != -1;
+}
+
+/** Close frame buffer device
+ */
+void mce_fbdev_close(void)
+{
+    if( mce_fbdev_handle == -1 )
+        goto EXIT;
+
+    mce_log(LL_NOTICE, "closing frame buffer device");
+    close(mce_fbdev_handle), mce_fbdev_handle = -1;
+    mce_log(LL_DEBUG, "closed frame buffer device");
+
+EXIT:
+    return;
+}
+
+/** Reopen frame buffer device unless denied
+ */
+void mce_fbdev_reopen(void)
+{
+    if( mce_fbdev_allow_open )
+        mce_fbdev_close();
+
+    mce_fbdev_open();
+}
+
+/* ========================================================================= *
+ * POST_EXIT_LINGER
+ * ========================================================================= */
+
+/** Create a child process to keep frame buffer device open after mce exits
+ *
+ * The frame buffer device powers off automatically when the last open
+ * file descriptor gets closed.
+ *
+ * To allow the shutdown logo to stay on screen after lipstick and mce
+ * have been terminated, we create a detached child process that hangs
+ * on to frame buffer device.
+ *
+ * @param delay_ms how long the child process should linger
+ */
+void mce_fbdev_linger_after_exit(int delay_ms)
+{
+    static const char msg[] = "closing frame buffer device after delay\n";
+
+    /* Fork a child process */
+
+    int child_pid = fork();
+
+    /* Deal with parent side and return to caller */
+
+    if( child_pid != 0 ) {
+        if( child_pid < 0 )
+            mce_log(LL_ERR, "forking fbdev linger child failed: %m");
+        else
+            mce_log(LL_DEBUG, "fbdev linger child: pid %d", child_pid);
+        return;
+    }
+
+    /* Detach from parent so that we will not get killed with it */
+
+    setsid();
+
+    /* Close all files, except fbdev & stderr */
+
+    int nfd = getdtablesize();
+    for( int fd = 0; fd < nfd; ++fd ) {
+        if( fd != mce_fbdev_handle && fd != STDERR_FILENO )
+            close(fd);
+    }
+
+    if( delay_ms < 500 )
+        delay_ms = 500;
+
+    /* Wait ... */
+
+    struct timespec ts =
+    {
+        .tv_sec  = (time_t)(delay_ms / 1000),
+        .tv_nsec = (long)(delay_ms % 1000) * 1000000,
+    };
+
+    while( nanosleep(&ts, &ts) == -1 && errno == EINTR ) { /* nop */ }
```

Comment:
No, this is normal way to meet expected sleep time. The 1st arg is time to sleep, the 2nd is updated by nanosleep to contain remaining time if it returns prematurely.

"EINTR  The  pause has been interrupted by a signal that was delivered to the thread.  The remaining sleep time has been written into *rem so that the thread can easily call nanosleep()again and continue with the pause."